packages/entity/src/EntityAssociationLoader.ts

```diff
@@ -54,14 +55,20 @@ export default class EntityAssociationLoader<
       .getViewerScopedEntityCompanionForClass(associatedEntityClass)
       .getQueryContextProvider()
       .getQueryContext()
-  ): Promise<Result<TAssociatedEntity>> {
+  ): Promise<
+    Result<null extends TFields[TIdentifyingField] ? TAssociatedEntity | null : TAssociatedEntity>
+  > {
     const associatedEntityID = this.entity.getField(fieldIdentifyingAssociatedEntity);
+    if (!associatedEntityID) {
+      return result(null) as any;
+    }
+
     const loader = this.entity
       .getViewerContext()
       .getViewerScopedEntityCompanionForClass(associatedEntityClass)
       .getLoaderFactory()
       .forLoad(queryContext);
-    return await loader.loadByIDAsync((associatedEntityID as unknown) as TAssociatedID);
+    return (await loader.loadByIDAsync((associatedEntityID as unknown) as TAssociatedID)) as any;
```

Comment:
This too - I'd like to better understand why we need the `any` cast and if some simple `readonly`s can fix it.

Reply:
Temporarily changing the return type of the function to `Result<TAssociatedEntity | null>` and returning `result(null)` indicates to me that it isn't the result library (which is covariant), but instead a loss of information about the conditional type. I may be misunderstanding your suggestion though.

Reply:
What about `Promise<Result<TAssociatedEntity> | Result<null>>`? My question about covariance is whether `const r: Result<TAssociatedEntity | null> = result(null)` passes. I believe a correct type checker should reject this unless `Result<T>` treats `T` as `readonly`.

Reply:
Changing the return type to that and returning `r` does trigger an error.
```
src/__tests__/EntityAssociationLoader-test.ts(26,14): error TS2531: Object is possibly 'null'.
src/EntityAssociationLoader.ts(62,7): error TS2322: Type 'Success<TAssociatedEntity | null>' is not assignable to type 'Success<TAssociatedEntity> | Failure<TAssociatedEntity> | Success<null> | Failure<null>'.
  Type 'Success<TAssociatedEntity | null>' is not assignable to type 'Success<TAssociatedEntity>'.
    Type 'TAssociatedEntity | null' is not assignable to type 'TAssociatedEntity'.
      'TAssociatedEntity' could be instantiated with an arbitrary type which could be unrelated to 'TAssociatedEntity | null'.
        Type 'null' is not assignable to type 'TAssociatedEntity'.
          'TAssociatedEntity' could be instantiated with an arbitrary type which could be unrelated to 'null'.
```
This error message makes sense to me, though I need to read up on how `readonly` is used in the type system since I'm not following that piece.

Reply:
This part of the error makes sense to me: `Type 'Success<TAssociatedEntity | null>' is not assignable to type 'Success<TAssociatedEntity>'.` We maybe should dig into that more.

I did some testing on the side and found this checks out `const r1: Result<string | null> = result(null) as Result<null>`. I find this surprising, as it holds even if the `readonly`s are removed from `Result`:

```
    const r0 = result(1);
    const r1: Result<string | number> = r0;
    r1.value = 'hi not a number';
```

Reply:
Lol the following works:

```typescript
return result(null) as Result<
  null extends TFields[TIdentifyingField] ? TAssociatedEntity | null : TAssociatedEntity
>;
```